src/hotspot/os_cpu/linux_aarch64/vm_version_linux_aarch64.cpp

```diff
@@ -167,3 +167,23 @@ void VM_Version::get_os_cpu_info() {
     fclose(f);
   }
 }
+
+void VM_Version::get_compatible_board(char *buf, int buflen) {
+  assert(buf != NULL, "invalid argument");
+  assert(buflen >= 1, "invalid argument");
+  *buf = '\0';
+  int fd = open("/proc/device-tree/compatible", O_RDONLY);
+  if (fd != -1) {
+    ssize_t read_sz = read(fd, buf, buflen);
+    buf[buflen - 1] = '\0';
+    if (read_sz != -1) {
+      // Replace '\0' to ' '
+      for (char *ch = buf; ch < buf + read_sz; ch++) {
+        if (*ch == '\0') {
+          *ch = ' ';
+        }
+      }
+    }
+    close(fd);
+  }
+}
```

Comment:
It's still wrong. If the read() call has filled the whole buffer, the string is not zero-terminated when this function returns.

Reply:
So I set `\0` to the tail of `buf` at L178.

Reply:
Thanks for your comment! I pushed new commit.

I issue `read()` with `buflen - 1` bytes, then I set `\0' to tail of `buf` after `read()`. And also `\0` will be converted to blank will happen when `read()` return 1 or greater. How about it?

Reply:
I think you need this to cope with errors, empty files, and so on:

```
void VM_Version::get_compatible_board(char *buf, int buflen) {
  assert(buf != NULL, "invalid argument");
  assert(buflen >= 1, "invalid argument");
  *buf = '\0';
  int fd = open("/proc/device-tree/compatible", O_RDONLY);
  ssize_t read_sz = read(fd, buf, buflen - 1);
  if (read_sz >= 0) {
    buf[read_sz] = '\0';
    // Replace '\0' to ' '                                                                                                                                                                                                                                                                   
    for (char *ch = buf; ch < buf + read_sz; ch++) {
      if (*ch == '\0') {
        *ch = ' ';
      }
    }
  } else { // read() retuned an error                                                                                                                                                                                                                                                        
    buf[0] = '\0';
  }
=>close(fd);
}
```

Reply:
I updated PR, but it is different from your suggestion a bit. This version would return empty string when `open()` or `read()` returns error, and also `close()` would be called when `open()` succeeded.

Did you say we can call `read()` and `close()` even if `open()` failed? According to manpage, it seems to work (just return EBADF), but I feel strange a bit. If it is ok, I will do so.

Reply:
No, if open() fails we should return straight away, with an empty string. That needs an addition.
We must, however, terminate the string with 0 at the correct point, at the end of the bytes read. Otherwise` strlen() `reads uninitialized memory.

Reply:
I think my latest commit includes your suggestion:

* returns empty string ( `\0` ) when `open()` failed.
* returns empty string when `read()` failed or read nothing (returns `0` )
* add `\0` to `buf[read_sz]` just after `read()` call, and skip it at the loop - it can be assumed `\0` is set to tail of `buf`

Or should I change as following for readability?

```
int fd = open("/proc/device-tree/compatible", O_RDONLY);
if (fd == -1) {
  *buf = '\0';
  return;
}

ssize_t read_sz = read(fd, buf, buflen - 1);
if (read_sz <= 0) {
  *buf = '\0';
  return;
}

// Add '\0' to the tail
buf[read_sz] = '\0';
// Replace '\0' to ' '
for (char *ch = buf; ch < buf + read_sz; ch++) {
  if (*ch == '\0') {
    *ch = ' ';
  }
}

close(fd);
```

Reply:
I tested current PR manually with this Gist code, it works fine without buffer overrun.
https://gist.github.com/YaSuenag/bdc73c3540335a096fa289ead36ca8b5

Reply:
That looks right.